### PR TITLE
draft of api design

### PR DESCRIPTION
**Pull request by ttdonovan**
rough-draft of api design - any feedback or suggested changes is welcomed. #2

### CODE REVIEW

File: docs/api_design.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    |-- Protocol
+        |-- `send()` and `receive()`
+|-- SC2API
+    |-- Request/Response
+        |-- (protocol types)
+```
+
+## Bot
+
+The agent used to infer **GameState** and take an **Action**. The **Bot** will
```
Comment (Michael-F-Bryan) [inline docs/api_design.md]: I'd probably say "inspect" instead of "infer" here.

File: docs/api_design.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+between the **Engine** and **Connection**.
+
+### Controller
+
+Primarily used to `act()` and `observe()`. Also will be able to make
+`ping()` and `game_info()` types of one-off requests.
+
+### Protocol
+
+Will send and receive **SC2API** request and response via a **Connection**.
```
Comment (Michael-F-Bryan) [inline docs/api_design.md]: Hmm, this sounds an awful lot like what the `Connection` will be doing. I reckon we can skip the `Protocol` step and the controller contains a `Connection` internally and does things like `self.conn.ping()` and `self.conn.send_action(some_action)`. Then the `Connection` constructs the corresponding protobuf messages internally, converting high level `Action`s into whatever, then it sends that to the websocket client over a channel.

File: docs/api_design.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+be notified of **GameState** via a `tick()`.
+
+### Action
+
+In-game actions that can be performed by a **Bot**.
+
+## Connection
+
+Responsible for establishing the WebSocket connection to a StarCraft II running
+process for sending and receiving messages. Used by the **Remote::Protocol**.
```
Comment (Michael-F-Bryan) [inline docs/api_design.md:41]: We may have an issue here in that there needs to be two-way communication between `Connection` and the websocket client. Therefore you're either going to need two channels, or will have to do something fancy with callbacks.

I'll have a dig through the C++ API and see how they deal with it...

File: docs/api_design.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+|-- Connection
+|-- Engine
+    |-- GameState
+    |-- `run()`
+|-- Remote
+    |-- Controller
+        |-- `act()`, `observe()`, `ping()`, ...
+    |-- Protocol
+        |-- `send()` and `receive()`
+|-- SC2API
```
Comment (Michael-F-Bryan) [inline docs/api_design.md]: I reckon a better name would be `raw_protobuf_api`, or something along those lines. Naming it *almost* the same as the crate will be confusing, plus we want to put emphasis on the fact that it's contains raw protobuf stuff and shouldn't be accessible by end users... Thoughts?

File: docs/api_design.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    |-- Request/Response
+        |-- (protocol types)
+```
+
+## Bot
+
+The agent used to infer **GameState** and take an **Action**. The **Bot** will
+be notified of **GameState** via a `tick()`.
```
Comment (Michael-F-Bryan) [inline docs/api_design.md]: How about replacing the second sentence with something like this:

> at every game frame, the `Bot`'s `tick()` method will be called to let the bot inspect the current game state and emit actions (if desired).

File: docs/api_design.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+### Action
+
+In-game actions that can be performed by a **Bot**.
```
Comment (Michael-F-Bryan) [inline docs/api_design.md]: A dumb `enum` of all the actions a `Bot` can emit.

**Review by Michael-F-Bryan — CHANGES_REQUESTED** [review]

File: docs/api_design.md (hunk 2 above)
Comment (ttdonovan) [reply docs/api_design.md]: It's not clear to me yet who is going to be responsible for the construction the protocol buffers. My initial though was that the`Connection` would have two public functions `send_request(request: raw_protobuf_api::Request)` and `resc_response() -> Result<raw_protobuf_api::Reponse>` and `Protocol` would be the intermediate layer between lib and WebSocket connection. The `Connection` is then only responsible for encoding/decoding messages from a WebSocket.

**Pull request merged**